### PR TITLE
feat: Support invoice line items (il_) from Stripe API 2019-12-03

### PR DESCRIPTION
**Pull request by bravier**
Stripe API is returning `InvoiceLineItem`s in its latest version
instead of `InvoiceItem`s as it may have in the past.

This new object acts as a representation of either an
`InvoiceItem` or a `SubscriptionItem` for each line of an `Invoice`.

More info on [Stripe's documentation](https://stripe.com/docs/api/invoices/line_item).

Let's support them in localstripe!

### CODE REVIEW

File: localstripe/resources.py [hunk 1]
```diff
@@ -1400,6 +1395,70 @@ def _api_list_all(cls, url, customer=None, limit=None):
...
+        self.metadata = item.metadata
+
+        self.amount = item.amount \
+            if self.type == 'invoiceitem' else item._calculate_amount()
+        self.currency = \
+            item.currency if self.type == 'invoiceitem' else item.plan.currency
+        self.description = \
+            item.description if self.type == 'invoiceitem' else item.plan.name
+        self.proration = \
+            item.proration if self.type == 'invoiceitem' else False
```
Comment (bravier) [inline localstripe/resources.py]: `proration` can only exist when the `InvoiceLineItem` references an `InvoiceItem`.
Using `False` as the default value here for `SubscriptionItem`s.

File: localstripe/resources.py [hunk 2]
```diff
@@ -1400,6 +1395,70 @@ def _api_list_all(cls, url, customer=None, limit=None):
...
+
+        self.quantity = item.quantity if self.type == 'subscription' else 1
```
Comment (bravier) [inline localstripe/resources.py]: In `localstripe` the quantity can only be set on `SubscriptionItem`s so I use `1` as the default value for `InvoiceItem`s.

File: localstripe/resources.py [hunk 3]
```diff
@@ -1378,11 +1378,6 @@ def __init__(self, invoice=None, subscription=None, plan=None, amount=None,
         self.tax_rates = tax_rates
         self.metadata = metadata or {}
 
-    @property
-    def tax_amounts(self):
-        if self.tax_rates is not None:
-            return [tr._tax_amount(self.amount) for tr in self.tax_rates]
-
```
Comment (adrienverge) [inline localstripe/resources.py]: :+1:

File: localstripe/resources.py [hunk 4]
```diff
@@ -2304,16 +2363,7 @@ def _create_invoice(self):
...
-                InvoiceItem(subscription=self.id, plan=si.plan.id,
-                            amount=si._calculate_amount(),
-                            currency=si.plan.currency,
-                            description=si.plan.name,
-                            tax_rates=[tr.id
-                                       for tr in (si.tax_rates or [])],
-                            customer=self.customer,
-                            period_start=self.current_period_start,
-                            period_end=self.current_period_end))
+            pending_items.append(si)
```
Comment (adrienverge) [inline localstripe/resources.py]: The code looks good, but can it be a problem to reuse the same objects here (instead of creating new objects, like the previous code did)? @H--o-l I think you reworked this code not so long ago (commit 5fca63eb), do you have an opinion?

File: localstripe/resources.py (hunk 2 above)
Comment (adrienverge) [reply localstripe/resources.py]: Perfect! It seems in agreement with what Stripe API doc examples show.

File: localstripe/resources.py [hunk 5]
```diff
@@ -1400,6 +1395,70 @@ def _api_list_all(cls, url, customer=None, limit=None):
...
+
+        self.type = \
+            'invoiceitem' if isinstance(item, InvoiceItem) else 'subscription'
+
+        if self.type == 'subscription':
+            self.subscription_item = item.id
+            self.subscription = item._subscription
+        elif self.type == 'invoiceitem':
+            self.invoice_item = item.id
+            self.subscription = None
```
Comment (adrienverge) [inline localstripe/resources.py]: Maybe I suggest the following? It seems more readable to me:
```python
        if self.type == 'subscription':
            self.subscription_item = item.id
            self.subscription = item._subscription
            self.plan = item.plan
            self.quantity = item.quantity
            self.amount = item._calculate_amount()
            self.currency = item.plan.currency
            ...
            subscription_obj = Subscription._api_retrieve(item._subscription)
            self.period = dict(start=subscription_obj.current_period_start,
                               end=subscription_obj.current_period_end)
        elif self.type == 'invoiceitem':
            self.invoice_item = item.id
            self.subscription = None
            self.plan = None
            self.quantity = 1
            self.currency = item.currency
            ...
            self.amount = item.amount
            self.period = item.period
```

File: localstripe/resources.py (hunk 2 above)
Comment (bravier) [reply localstripe/resources.py]: Please note that `InvoiceItem` has a `quantity` field (see the [doc](https://stripe.com/docs/api/invoiceitems/object#invoiceitem_object-quantity)) but localstripe do not implement it. Maybe a comment in the code can help better understand this default value?

Comment (adrienverge) [reply localstripe/resources.py]: Ah, thanks, I hadn't seen that!

Maybe a semi-quick-win would be to add this in `InvoiceItem`:
```diff
@@ -1368,6 +1368,7 @@ class InvoiceItem(StripeObject):
         self.invoice = invoice
         self.subscription = subscription
         self.plan = plan
+        self.quantity = 1
         self.amount = amount
         self.currency = currency
         self.customer = customer
```
and just use this in `InvoiceLineItem`:
```python
self.quantity = item.quantity
```
?

File: localstripe/resources.py (hunk 5 above)
Comment (bravier) [reply localstripe/resources.py]: Indeed it is. Done!

File: localstripe/resources.py (hunk 2 above)
Comment (bravier) [reply localstripe/resources.py]: Good idea. It would be clearer on why this number is 1 by default. Code updated.

File: localstripe/resources.py (hunk 4 above)
Comment (H--o-l) [reply localstripe/resources.py]: Hey!

For me it looks OK, because a new object is not created here, but will be at this line https://github.com/adrienverge/localstripe/pull/119/files#diff-62ee820ebbbef6539c97fecb3a10c374R915

Object that are created and should not be removed without breaking something are :
https://github.com/adrienverge/localstripe/blob/c6821928953c50058327f04ef11b9e70bc8e609b/localstripe/resources.py#L2471-L2476
These end-ups being the one we get with:
```py
        pending_items = [ii for ii in InvoiceItem._api_list_all(
            None, customer=customer, limit=99)._list
            if ii.invoice is None]
```
TL;DR: OK for me! @baptsite as said on Slack, unit test could be cool, and the following API https://stripe.com/docs/api/invoices/invoice_lines too otherwise OK for me.

Comment (adrienverge) [reply localstripe/resources.py]: @H--o-l thanks :+1:

**Comment by bravier** [comment]
Hi Adrien,
Small update to let you know I've added support for retrieving an invoice's line item on a specific route (https://stripe.com/docs/api/invoices/invoice_lines) and a small test to verify this route exists on an invoice.
I noticed parameters `ending_before` and `starting_after` are not implemented for other routes which might support them so I didn't implemented them here neither. Hope it's ok like that.

File: localstripe/resources.py [hunk 6]
```diff
@@ -1304,11 +1304,24 @@ def _api_void_invoice(cls, id):
...
 
+    @classmethod
+    def _api_list_lines(cls, id, limit=None, **kwargs):
+        if kwargs:
+            raise UserError(400, 'Unexpected ' + ', '.join(kwargs.keys()))
+
+        obj = cls._api_retrieve(id)
+
+        lines = List('/v1/invoices/' + id + '/lines', limit=limit)
+        lines._list = obj.lines._list
```
Comment (adrienverge) [inline localstripe/resources.py]: I'd be something simpler, like in
https://github.com/adrienverge/localstripe/blob/c6821928953c50058327f04ef11b9e70bc8e609b/localstripe/resources.py#L749-L753
or
https://github.com/adrienverge/localstripe/blob/c6821928953c50058327f04ef11b9e70bc8e609b/localstripe/resources.py#L741-L746

Is it possible, or is it a hidden reason for the need to create a new `List`?

File: test.sh [hunk 7]
```diff
@@ -399,11 +399,12 @@ curl -sSf -u $SK: $HOST/v1/subscriptions \
...
-           -d customer=$cus \
-           -d items[0][plan]=basique-mensuel \
-           -d items[0][tax_rates][0]=$txr1 \
-      | grep -oE 'sub_\w+' | head -n 1)
+res=$(curl -sSf -u $SK: $HOST/v1/subscriptions \
+                    -d customer=$cus \
+                    -d items[0][plan]=basique-mensuel \
+                    -d items[0][tax_rates][0]=$txr1)
+sub=$(echo "$res" | grep -oE 'sub_\w+' | head -n 1)
+inv=$(echo "$res" | grep -oE 'in_\w+' | head -n 1)
```
Comment (adrienverge) [inline test.sh]: This is nitpicking, but can you use `in`, to stay within Stripe prefixes?

File: localstripe/resources.py (hunk 6 above)
Comment (bravier) [reply localstripe/resources.py]: It was mainly to make use of the `List` class validation code for the `limit` parameter instead of copying it.

Comment (adrienverge) [reply localstripe/resources.py]: Ah, OK! It's a good reason.

File: test.sh (hunk 7 above)
Comment (bravier) [reply test.sh]: Code updated:
```diff
--- a/test.sh
+++ b/test.sh
@@ -404,7 +404,7 @@ res=$(curl -sSf -u $SK: $HOST/v1/subscriptions \
                     -d items[0][plan]=basique-mensuel \
                     -d items[0][tax_rates][0]=$txr1)
 sub=$(echo "$res" | grep -oE 'sub_\w+' | head -n 1)
-inv=$(echo "$res" | grep -oE 'in_\w+' | head -n 1)
+in=$(echo "$res" | grep -oE 'in_\w+' | head -n 1)
 
 curl -sSf -u $SK: $HOST/v1/invoices?customer=$cus
 
@@ -414,7 +414,7 @@ curl -sSf -u $SK: $HOST/v1/invoices/upcoming?customer=$cus\&subscription_items%5
 
 curl -sSf -u $SK: $HOST/v1/invoices/upcoming?customer=$cus\&subscription=$sub\&subscription_items%5B0%5D%5Bid%5D=si_RBrVStcKDimMnp\&subscription_items%5B0%5D%5Bplan%5D=basique-annuel\&subscription_proration_date=1504182686\&subscription_tax_percent=20
 
-curl -sSf -u $SK: $HOST/v1/invoices/$inv/lines
+curl -sSf -u $SK: $HOST/v1/invoices/$in/lines
 
 cus=$(curl -sSf -u $SK: $HOST/v1/customers \
            -d description='This customer will have a subscription with volume tiered pricing' \
```

**Comment by adrienverge** [comment]
Perfect!
